File: 0238-product-of-array-except-self/0238-product-of-array-except-self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function(nums) {
    let n = nums.length;
    let pre = [];
    let suf = [];
    pre[0] = 1;
    suf[n-1] = 1;
    for(let i=1;i<n;i++){
        pre[i] = pre[i-1]*nums[i-1];
    }
    for(let i=n-2;i>=0;i--){
        suf[i] = suf[i+1]*nums[i+1];
    }
    for(let i=0;i<n;i++){
        nums[i] = pre[i]*suf[i];
    }
    return nums;
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] prefix = new int[n];
        int[] suffix = new int[n];
        prefix[0] = 1;
        suffix[n-1] = 1;
        for(int i=1;i<n;i++){
            prefix[i] = prefix[i-1]*nums[i-1];
        }
        for(int i=n-2;i>=0;i--){
            suffix[i] = suffix[i+1]*nums[i+1];
        }
        for(int i=0;i<n;i++){
            nums[i] = prefix[i] * suffix[i];
        }
        return nums;
    }
}